### PR TITLE
Spice for keybase.io

### PR DESCRIPTION
**Pull request by mwolny**
**What does your Instant Answer do?**
It retrieves basic information about user from keybase.io.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Shoot me I do not know how to answer to this question.

**What is the data source for your Instant Answer? (Provide a link if possible)**
[keybase.io API](https://keybase.io/docs/api/1.0)

**Why did you choose this data source?**
I like [keybase.io](https://keybase.io) and you do not have any instant answer for it.

**Are there any other alternative (better) data sources?**
Not to my knowledge.

**What are some example queries that trigger this Instant Answer?**
```keybase mwolny``` ```keybase jeresig```

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Well I strong believe that's geek only goodie.

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
It's not.

**Which existing Instant Answers will this one supersede/overlap with?**
None. 

**Are you having any problems? Do you need our help with anything?**
Nope. 

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
API is open for everyone, take a look at [API docs](https://keybase.io/docs/api/1.0), on the top. 

**Where did you hear about DuckDuckHack? (For first time contributors)**
Well I was reading about [duckduckgo](https://duckduckgo) and at some point I found [duckduckhack](http://duckduckhack.com).

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
Example screenshot

### CODE REVIEW

**Comment by MrChrisW** [comment]
@mwolny You're right, I definitely agree that providing the pub key makes this IA useful. 

That sounds like a good idea. However we generally try to limit the number of requests an IA makes so we may want to limit the number of users that are returned. 

@mattr555 @moollaza @jagtalon 
What do you think about displaying multiple users like the *Google +* IA or the *US Senators* IA for this IA? Or is displaying a single user best? Thanks for your input!

**Comment by mwolny** [comment]
Yep, I can imagine that request number is an important factor but in this case second request is done only when user click to get details about specific user from the lookup result.

Anyway I will for the input from the rest of the team.

**Comment by mattr555** [comment]
@MrChrisW @mwolny Meh, I don't think it's that important to display multiple results. Keybase works around the idea that you know who you're communicating with. I wouldn't, for example, be looking up Keybase users named "Dave" to talk to.

It may be more interesting to be able to look up users by their other account integrations. For example, if you could (theoretically, as I don't have a Keybase account) look me up with `keybase twitter:mattr555` or `keybase github:mattr555`, that could be pretty useful.

Just my $.02

**Comment by jagtalon** [comment]
@mwolny awesome! I'm not a fan of just showing the GPG key, though. I think the Twitter look is perfect for this though! https://duckduckgo.com/?q=%40duckduckgo&ia=social

screen shot 2015-01-01 at 2 09 21 pm

It's very similar to what they show on Keybase.io:

screen shot 2015-01-01 at 2 09 48 pm

I think it's more pleasing to the eyes. What do you think @mattr555 @MrChrisW?

**Comment by jagtalon** [comment]
I agree with @mattr555--no need for multiple results.

**Comment by mwolny** [comment]
What do you guys think about it?

**Comment by mwolny** [comment]
One thing I forgot to say. These key fingerprints are links to ``https://keybase.io/<user>/key.asc``.

**Comment by mwolny** [comment]
Example 3

**Comment by jagtalon** [comment]
@mwolny Nice--would like the image to be round (see twitter code https://github.com/duckduckgo/zeroclickinfo-spice/tree/master/share/spice/twitter) and I think the bottom should be just the description instead of the fingerprint:

screen shot 2015-01-01 at 7 18 45 pm

**Comment by mwolny** [comment]
Well I decided to skip description because a very small amount of user has it or I had bad luck. :-)
Check this out:
Example jeresig
Example jbx

If you good with it I will commit my changes.

**Comment by MrChrisW** [comment]
@mwolny Definitely push your commits so the code can be reviewed. I'm liking this design :grinning: 

The only issue is the spacing when no description is found:
selection_038

**Comment by mwolny** [comment]
Example jbx
Example jeresig

**Comment by mwolny** [comment]
**What does your Instant Answer do?**
It retrieves basic information about user from keybase.io.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Shoot me I do not know how to answer to this question.

**What is the data source for your Instant Answer? (Provide a link if possible)**
[keybase.io API](https://keybase.io/docs/api/1.0)

**Why did you choose this data source?**
I like [keybase.io](https://keybase.io) and you do not have any instant answer for it.

**Are there any other alternative (better) data sources?**
Not to my knowledge.

**What are some example queries that trigger this Instant Answer?**
```keybase mwolny``` ```keybase jeresig```

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Well I strong believe that's geek only goodie.

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
It's not.

**Which existing Instant Answers will this one supersede/overlap with?**
None. 

**Are you having any problems? Do you need our help with anything?**
Nope. 

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
API is open for everyone, take a look at [API docs](https://keybase.io/docs/api/1.0), on the top. 

**Where did you hear about DuckDuckHack? (For first time contributors)**
Well I was reading about [duckduckgo](https://duckduckgo) and at some point I found [duckduckhack](http://duckduckhack.com).

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
Example screenshot

**Comment by MrChrisW** [comment]
@mwolny Thanks for making those changes! Looks great so far :grinning: I've added the template to your first comment.

**Comment by jagtalon** [comment]
@mwolny @MrChrisW Yup, I definitely like those better!

File: share/spice/keybase/keybase.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            data: user,
+            meta: {
+                sourceName: 'keybase.io',
+                sourceUrl: 'https://keybase.io/' + user.basics.username
+            },
+            templates: {
+                group: 'base',
+                options: {
+                    content: Spice.keybase.content,
+                    moreAt: true,
```
Comment (jagtalon) [inline share/spice/keybase/keybase.js]: Remove trailing comma--this fails in IE 8.

**Comment by mwolny** [comment]
Good to hear it. :-) Thanks.
@jagtalon: just uploaded a fix

**Comment by jagtalon** [comment]
@mwolny Sweet! Taking a quick look

**Comment by MrChrisW** [comment]
Mobile layout.
Twitter:
tmp_14149-_20150106_052958836027846

Keybase
tmp_14149-_20150106_053016-442943690


Looks like the description text needs to be smaller and the name bold.

**Comment by jagtalon** [comment]
Nice find @MrChrisW

**Comment by mwolny** [comment]
I will try to fix it tomorrow :-) Great catch!

**Comment by mwolny** [comment]
Fixed. I changed also displaying of user full name. Now it looks like Twitter spice.

**Comment by jagtalon** [comment]
@mwolny Nice

@abeyang it's up on https://ddh1.duckduckgo.com/?q=keybase+jeresig&ia=keybase!

**Comment by mattr555** [comment]
> For example, if you could ~~(theoretically, as I don't have a Keybase account)~~ _got one as of today_ look me up with `keybase twitter:mattr555` or `keybase github:mattr555`, that could be pretty useful.

Is this going to happen? Don't mean to feature creep, but it's possible to do [right from the same api endpoint](https://keybase.io/docs/api/1.0/call/user/lookup).

**Comment by MrChrisW** [comment]
@mattr555 You're such a feature creep! :laughing: (wasn't aware of this term)

This may be possible, however I don't think the `spice to` URL parms can be updated based on query input? Tell me if I'm wrong. Could make a new module for each social parm (twitter, github) and hit that from the JavaScript? 

@mwolny Can you add the fields parameter to the API URL as suggested in my line comments? Thanks!

**Comment by mattr555** [comment]
@MrChrisW @mwolny We could use [`spice from`](https://duck.co/duckduckhack/spice_advanced_backend#returning-multiple-values-to-spice-from). Basically we would create a regex to insert the values in like so:

```perl
spice to => 'https://keybase.io/_/api/1.0/user/lookup.json?fields=basics,public_keys&$1'
spice from => '(.*)=(.*)'
```

Then just parse out the service before and after the colon and return the values like `return $service, $username` (defining these variables, of course). When there's no defined service, assume they want the keybase username: `return 'usernames', $username`

And yeah, [feature creep](https://en.wikipedia.org/wiki/Feature_creep) is when too many features are added at the expense of simple design :neckbeard:

**Comment by mwolny** [comment]
I'm missing one information. Even if we add ability to lookup users with ``keybase github:mattr555`` or ```twitter:mattr555``` patterns.. how does duckduckgo user  get to know about it?

**Comment by MrChrisW** [comment]
@mwolny We can display tha under "Example" queries on the Keybase IA information page on duck.co. See the [editor IA page](https://duck.co/ia/view/editor).

Simply add it as a secondary query in your Perl. :+1:  Let me know if you have any issues with implementation.

**Comment by mwolny** [comment]
Gotcha! Gonna do it tomorrow during the day. :-) Thanks!

**Comment by mwolny** [comment]
Well I did changes but I have a problem. When I use ``keybase github:mwolny`` it tries to use Github spice instead of keybase. Pretty the same thing with twitter and reddit. Could you point me how to solve it?

Whats more if you do not define the service name it will fallback to ``username``. It means that you can use this spice ``keybase mwolny`` and ``keybase username:mwolny``.

**Comment by mwolny** [comment]
Forgot to ask. Could some of you explain me how does ``spice from`` work? When I tried to use ``spice from => '(\w+):(\w+)'``` it didn't work. When I changed it to ``spice from => '(\w+)/(\w+)'`` it started to work.

**Comment by mattr555** [comment]
@mwolny Sorry, my initial explanation for `spice from` was pretty wrong. When you return multiple values in your `handle` function, they will be joined with slashes. This string is then passed to the `from` regex, where you parse out the two groups to insert into the `to` string.

> Whats more if you do not define the service name it will fallback to `username`. It means that you can use this spice `keybase mwolny` and `keybase username:mwolny`.

Great! That's what I intended when I suggested this feature :smiley:

> Well I did changes but I have a problem. When I use ``keybase github:mwolny`` it tries to use Github spice instead of keybase. Pretty the same thing with twitter and reddit. Could you point me how to solve it?

I can't reproduce this...but I found that the service name lookup doesn't work. It turns out that `api_result.them` can be an array for some types of lookups, so you'll need to choose probably the first item.

**Comment by mwolny** [comment]
Ochhh, But if you use ``username`` ``api_result`` won't be an array. Well, I will add code that checks if ``api_result.them`` is an array. If so I will be using first item as you said and actually it was at the begining.

File: t/Keybase.t [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+use warnings;
+use Test::More;
+use DDG::Test::Spice;
+
+spice is_cached => 1;
+
+ddg_spice_test(
+    [qw( DDG::Spice::Keybase)],
+    'keybase jbx' => test_spice(
+        '/js/spice/keybase/jbx',
```
Comment (mattr555) [inline t/Keybase.t]: @mwolny Also be sure to update your tests and add some more for the different services.

Comment (mwolny) [inline t/Keybase.t]: Good point! Will do.

**Comment by mwolny** [comment]
[Here is an example](http://i.imgur.com/5Vyw5CD.png) when I tried to use ``keybase reddit:jeresig`` and ``reddit`` spice was invoked.

**Comment by MrChrisW** [comment]
@mwolny jeresig doesn't have his reddit linked to keybase so that query returns nothing from the API. 

The other problem is the array issue you mentioned above which is causing the line `                sourceUrl: 'https://keybase.io/' + user.basics.username` to fail, as `user` is an array.
            
To avoid the other social IA taking over the Keybase IA, you can add `signal: 'high'` as an option in `Spice.add({`

**Comment by mwolny** [comment]
@MrChrisW  yep, i know. I have a fix for js. Thanks for tip with Keybase IA.

File: share/spice/keybase/keybase.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+(function (env) {
+    'use strict';
+    env.ddg_spice_keybase = function(api_result){
+
+        if (api_result.error || !api_result) {
```
Comment (jagtalon) [inline share/spice/keybase/keybase.js]: `!api_result` should come first.

**Comment by mwolny** [comment]
@jagtalon as you wish :)

**Comment by jagtalon** [comment]
@mwolny thanks! Everyone, here's the latest code: https://ddh1.duckduckgo.com/?q=keybase+jeresig&ia=keybase

**Comment by MrChrisW** [comment]
@mwolny Looking good on mobile now! :+1:

File: share/spice/keybase/content.handlebars [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<div class="keybase--basic">
+  <div class="keybase--image--wrap">
+    {{#if pictures.primary.url}}
+    <img src="{{imageProxy pictures.primary.url}}" class="circle keybase--image">
+    {{else}}
+    <img src="{{imageProxy 'https://keybase.io/images/no_photo.png'}}" class="circle keybase--image">
+    {{/if}}
+  </div>
+  <div class="keybase--misc">
+    <h1 class="zci__header">{{profile.full_name}}</h1>
```
Comment (MrChrisW) [inline share/spice/keybase/content.handlebars]: Should this be a hyperlink? Like the [twitter](https://github.com/duckduckgo/zeroclickinfo-spice/blob/6a878b223c89a6f7b8de556bf7803030f6a9e8a7/share/spice/twitter/content.handlebars#L8) IA? @jagtalon

Comment (jagtalon) [inline share/spice/keybase/content.handlebars]: Yup should be a hyperlink!

**Comment by jagtalon** [comment]
:+1:

**Comment by jagtalon** [comment]
@mwolny it's live! https://duckduckgo.com/?q=keybase+jeresig&ia=keybase

**Comment by mwolny** [comment]
Yeah! really happy to see it!

**Pull request merged**